### ISSUE TITLE
The cache file conversion LESS to CSS ignores changes in files that have been announced @import

### ISSUE THREAD

**Issue by DsXack** [yiisoft/yii2#608]
Check of existence is the file cache in AssetConverter

```
public function convert($asset, $basePath)

            if (@filemtime("$basePath/$result") < filemtime("$basePath/$asset")) {
```

For example, I point to the Asset Bundle file bootstrap.less, in which imported several LESS files. After changing one of the imported LESS files will not re-conversion bootstrap.less

I solved this problem in yii1 that specifies the directories in which you want to track filemtime to each file compiled LESS.

**Comment by brandonkelly** [comment]
We just came up with a solution for this for Craft. We’re overriding AssetManager::publishDirectory() with the following method:

	protected function publishDirectory($src, $options)
	{
		// See if any of the subdirectories have a more recent modify date than $src
		$srcModTime = filemtime($src);
		$subdirs = glob($src.DIRECTORY_SEPARATOR.'*', GLOB_NOSORT | GLOB_ONLYDIR);

		foreach ($subdirs as $dir)
		{
			if (filemtime($dir) > $srcModTime)
			{
				@touch($src);
				break;
			}
		}

		return parent::publishDirectory($src, $options);
	}

So basically if any subdirectories have a more recent mtime than `$src`, we call `touch()` to bump `$src`’s mtime. Then we let Yii’s parent method take it from there.

Seems to be working quite well, and shouldn’t have too much of a performance hit.

**Comment by schmunk42** [comment]
For reference [`forceCopy` per AssetBunde](http://www.yiiframework.com/wiki/789/forcecopy-per-asset-bundle) - Yii Wiki

**Comment by Kolyunya** [comment]
Setting the `forceCopy` option to `true` makes working with `less` nearly impossible. Each page load takes like half a minute with couple of tens of `less` files. Wouldn't it be faster to check `@imports` recursively?

**Comment by klimov-paul** [comment]
This issue can be solved providing custom check for whether asset is outdated  or not.
In my project I have created a special AssetConverter for that:

```php
class AssetConverter extends \yii\web\AssetConverter
{
    /**
     * @var boolean whether to check if already converted assets are outdated (source has been modified since conversion).
     * While enabled this option will force re-conversion of the outdated assets, but will consume extra time.
     */
    public $checkOutdated = true;


    /**
     * @inheritdoc
     */
    public function convert($asset, $basePath)
    {
        $pos = strrpos($asset, '.');
        if ($pos !== false) {
            $srcExt = substr($asset, $pos + 1);
            if (isset($this->commands[$srcExt])) {
                list ($ext, $command) = $this->commands[$srcExt];
                $result = substr($asset, 0, $pos + 1) . $ext;
                if ($this->forceConvert || $this->isOutdated($basePath, $asset, $result, $srcExt)) {
                    $this->runCommand($command, $basePath, $asset, $result);
                }

                return $result;
            }
        }

        return $asset;
    }

    /**
     * @param string $basePath the directory the $asset is relative to.
     * @param string $asset the asset file path, relative to $basePath
     * @param string $result the converted asset file path, relative to $basePath
     * @param string $sourceExtension source asset file extension.
     * @return boolean whether asset is outdated or not.
     */
    protected function isOutdated($basePath, $asset, $result, $sourceExtension)
    {
        $resultModificationTime = @filemtime("$basePath/$result");
        if ($resultModificationTime === false || $resultModificationTime === null) {
            return true;
        }
        if (!$this->checkOutdated) {
            return false;
        }
        if ($resultModificationTime < @filemtime("$basePath/$asset")) {
            return true;
        }
        foreach (FileHelper::findFiles($basePath, ['only' => ["*.{$sourceExtension}"]]) as $filename) {
            if ($resultModificationTime < @filemtime($filename)) {
                return true;
            }
        }
        return false;
    }
}
```

**Comment by Kolyunya** [comment]
@klimov-paul thank you! It would be very useful to have something like this in core though.

**Comment by schmunk42** [comment]
We've created https://github.com/dmstr/yii2-web/blob/master/src/AssetBundle.php to overcome this limitation.

Von meinem iPhone gesendet

**Comment by Kolyunya** [comment]
@qiangxue @klimov-paul @samdark @cebe I can't quite get it why this issue is closed. Am I missing something or the asset converter is still broken?

If you use default asset converter and set `forceConvert` then every page load will take like a minute with a large number of files. If you don't set `forceConvert` then imported assets just won't recompile. So one just can't use default asset converter.

Basically you have two options - either implement some custom asset converter or just get rid of it completely and user some external tools like `watch`. The funny thing is that the official framework site [uses](https://github.com/yiisoft-contrib/yiiframework.com/blob/master/gulpfile.js#L99) the latter.

Can we have some _working_ asset converter in framework core? If no then what is the point of having [this section](http://www.yiiframework.com/doc-2.0/guide-structure-assets.html#asset-conversion) if the official guide?

**Comment by cebe** [comment]
@Kolyunya I am using the forceConvert on small sites. If you whole compilation step takes a minute that is not a limitation of yii, but of the less compiler and you should consider using a different approach. There is seldomly one way that solves all problems. The auto conversion is for small sites, if you have a large set of less files you probably need to use other tools for it.

**Comment by schmunk42** [comment]
In my examples above we were trying just to look for modification times of files in the the asset folders. We're setting the modification time of the folder to the time of the newest file.
So we don't have to `forceCopy` all the time, but only when a modification time changes (by changing the hash), which is much faster.

What's about  a `deepHash` option, which could be turned on during development?

**Comment by cebe** [comment]
I see situation can be improved, so I'm reopening this. How fast is PHP on iterating over directories and getting mtimes? If that works fast enough, we could implement something that uses such an approach.

**Comment by cebe** [comment]
wait, this seems to be duplicate with https://github.com/yiisoft/yii2/issues/8197
